crates/storage/src/collections/mapping/mod.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A simple mapping to contract storage.
+
+use crate::traits::{
+    push_spread_root,
+    pull_spread_root,
+    clear_spread_root,
+    ExtKeyPtr,
+    KeyPtr,
+    SpreadLayout,
+};
+use ink_env::hash::{
+    Blake2x256,
+    HashOutput,
+};
+use ink_primitives::Key;
+
+use core::marker::PhantomData;
+
+/// A mapping of key-value pairs directly into contract storage.
+///
+/// If a key does not exist the `Default` value for the `value` will be returned.
+pub struct Mapping<K, V> {
+    key: Key,
+    _phantom_mapping: (PhantomData<K>, PhantomData<V>),
+}
```

Comment:
With `Option<Key>` we'd have a way to store information as to how the `Mapping` got initialized. It would be `None` in an `#[ink(constructor)]` called with `Mapping::new()` and it would be `Some(key)` if constructed with `SpreadLayout::pull_spread`.
The only downside to this approach is that this would make it impossible to use the `Mapping` within a constructor call.
For that exact reason we are caching the stored elements so that we can later push them down the sink.
Maybe we could do the same if we feel that this change is needed.

```rust
pub enum Mapping {
    New {
        ops: Vec<Operation<K, V>>,
    },
    Pulled {
        key: Key,
        _marker: PhantomData<fn() -> (K,V)>,
    }
}

pub enum Operation<K, V> {
    Insert(K, V),
    Erase(K),
}
```
However, this would no longer be as leightweight ... :sweat_smile:

Reply:
Why are you suggesting `fn() -> (K, V)` over just `(K, V)`?

>  The only downside to this approach is that this would make it impossible to use the Mapping within a constructor call.

Can you elaborate on this?

> For that exact reason we are caching the stored elements so that we can later push them down the sink.
Maybe we could do the same if we feel that this change is needed.

I agree, this wouldn't be lightweight and kinda defeats the point of the `Mapping`